content/en/BestPractices/Insecure_Deserialization.md

```diff
@@ -0,0 +1,296 @@
+
+if __name__ == '__main__':
+    app.run()
+
+```
+
+The server behaves in a similar fashion to the NodeJS example. It listens on port 5000. When a POST request is made to the endpoint, the server tries to decode and deserialize the data to extract a username.
+
+If there are errors, the server responds with a generic message otherwise replies with the name of the user. 
+
+Again, the serialized data is read directly from the request body without any checks. Therefore, it can be exploited easily to gain Remote code execution by generating payloads with the following script:
+
+```python
+import pickle
+import base64
+import os
+
+RCE_CMD = 'touch /tmp/PoC'
+
+class RCE:
+    def __reduce__(self):
+        cmd = (RCE_CMD)
+        return os.system, (cmd,)
+
+
+if __name__ == '__main__':
+    pickled = pickle.dumps(RCE())
+    print(base64.urlsafe_b64encode(pickled))
+```
+
+The output, when sent in the POST, executes the noted command on the system.
+
+```bash
+$ curl -X POST http://127.0.0.1:5000/ -d "{'payload': 'gASVKQAAAAAAAACMBXBvc2l4lIwGc3lzdGVtlJOUjA50b3VjaCAvdG1wL1BvQ5SFlFKULg=='}"
+```
+
+## Solution
+
+- Do not directly deserialize any user-controlled or untrusted data. For example, take the original code snippet:    
+
+```python
+class ExampleProtocol(protocol.Protocol):  
+	def dataReceived(self, data):    
+		# Code that would be here would parse the incoming data  
+		# After receiving headers, call confirmAuth() to authenticate  
+
+	def confirmAuth(self, headers):  
+		try:  
+			token = cPickle.loads(base64.b64decode(headers['AuthToken']))  # User control data flows into insecure cPickle.loads() function
+			if not check_hmac(token['signature'], token['data'], getSecretKey()):  
+				raise AuthFail  
+			self.secure_data = token['data']  
+		except:  
+			raise AuthFail
+```
+
+In the above example, data from request headers flow directly into the deserialization function, making it vulnerable. 
+
+-   Run untrusted deserialization function inside a sandbox. For example, you can use [nsjail](https://github.com/google/nsjail).
+-   Use deserialization methods like JSON, XML, and YAML that are language-agnostic.
+-   Use updated libraries, to avoid known security bugs.
+-   Limit and check all incoming and outgoing network activity from deserialization containers and servers.
+
+If you're a Java developer, you may need to implement the `serializable` interface due to inheritance. You can override the `readObject()` to prevent actual deserialization. Take the following example from [Snyk](https://snyk.io/)::
+
+```java
+private final void readObject(ObjectInputStream in) throws java.io.IOException { 
+	throw new java.io.IOException("Deserialized not allowed"); 
+}
+```
+- If you're using PHP, you may need to pass serialized data to the user. In that case, don't use the `unserialize()` function. Use a safe standard data interchange format such as `JSON` (via `json_decode()` and `json_encode()`).
```

Comment:
Do you have any reference information? 

You'll see in our other [Best Practices](https://developer.cobalt.io/bestpractices/) articles, others have provided appropriate references. I'd start with the relevant OWASP 2021 Top 10 entry.

Reply:
Hi @mjang-cobalt, as always, thanks for the feedback. I have made the changes, added the references and so on. I will be waiting for your feedback. Thank you for all the help and feedback 😀